numpy/core/shape_base.py

```diff
@@ -417,38 +419,99 @@ def format_index(index):
+                shape[axis+1:] != first_shape[axis+1:]):
+            raise ValueError('Mismatched array shapes in block.')
+    shape_on_dim = sum(shape[axis] for shape in shapes)
+    return first_shape[:axis] + (shape_on_dim,) + first_shape[axis+1:]
+
+
+def _atleast_nd(a, ndim):
+    # Ensures `a` has at least `ndim` dimensions by prepending
+    # ones to `a.shape` as necessary
+    return array(a, ndmin=ndim, copy=False, subok=True)
+
+
+def _accumulate(values):
+    # Helper function because Python 2.7 doesn't have
+    # itertools.accumulate
+    value = 0
+    accumulated = []
+    for v in values:
+        value += v
+        accumulated.append(value)
 
-    return block_recursion(arrays)
+    return tuple(accumulated)
+
+
+def _block_info_recursion(arrays, list_ndim, result_ndim, depth=0):
+    """Form required information to block the matricies using a flat iterator.
+
+    It returns a tuple containing
+
+    ``(shape, the_slices, the_arrays)``
+
+    * the resulting shape of the final array
+    * A flat iterable containing slices into the full array
+      required for assignment
+    * A flat iterable of arrays that need to be sliced in.
+
+    """
+    if depth < list_ndim:
+        axis = result_ndim - list_ndim + depth
+        info = [_block_info_recursion(arr, list_ndim, result_ndim, depth + 1)
+                for arr in arrays]
+        shapes, slices, arrays = zip(*info)
+        # Compute the offset for each array
+        offsets = (0,) + tuple(_accumulate(shape[axis]
+                                           for shape in shapes[:-1]))
+        # We contain lists that have the information from the lower dimensions
+        # To create the correct offset, we need to prepend the appropriate
+        # offset to them. Each list (containing a few different arrays)
+        # will require the matching offset computed above (with accumulate)
+        slices = [
+            [
+                (slice(offset, offset+shape[axis]),) + the_slice
+                for the_slice in inner_slices
+            ]
+            for offset, shape, inner_slices in zip(offsets, shapes, slices)
+        ]
+        shape = _concatenate_shapes(shapes, axis)
+
+        # Flatten the slices and arrays
+        slices = list(chain.from_iterable(slices))
+        arrays = list(chain.from_iterable(arrays))
+        if depth == 0:
+            # To correctly slice into the right dimention, we may need to
+            # prepend a few indices to the matrix.
+            slices = [(Ellipsis,) + the_slice for the_slice in slices]
+        return shape, slices, arrays
+    else:
+        # Base case
+        # Scalars have shape ()
+        shape = getattr(arrays, 'shape', ())
+        # broadcast the shape to the total number of dimensions
+        shape = (1,) * (result_ndim - len(shape)) + shape
+        # Return the slice and the array inside a list to be consistent
+        # with other case (i.e. not the base case)
```

Comment:
"the recursive case" is the better way of saying "[the] other case (i.e. not the base case)"

Reply:
ok.